### ISSUE TITLE
[CachePlugin] Streams are not cached

### ISSUE THREAD

**Issue by Nyholm** [php-http/plugins#20]
I found an issue with the CachePlugin. When we retrieve a `Response` form the cache we have lost the body because the stream was not in cache. 

``` php
if ($response->getBody()->__toString() === '') {
  // Always true
}
```

Im not sure how to solve this in a good way... One solution is to recreate the stream when we fetch the data from cache. This require us to add a dependency on `php-http/discovery`.

``` php
// If we can cache the request
$key = $this->createCacheKey($request);
$cacheItem = $this->pool->getItem($key);

if ($cacheItem->isHit()) {
     // return cached response
     $data = $cacheItem->get();
     $response = $data['response'];
     $response->withBody(StreamFactoryDiscovery::find()->createStream($data['body']));
     return new FulfilledPromise($response);
}

return $next($request)->then(function (ResponseInterface $response) use ($cacheItem) {
     if ($this->isCacheable($response)) {
         $cacheItem->set(['response'=>$response, 'body'=>$response->getBody()->__toString()])
              ->expiresAfter($this->getMaxAge($response));
              $this->pool->save($cacheItem);
      }

      return $response;
});

```

**Comment by sagikazarmark** [comment]
> This require us to add a dependency on `php-http/discovery`.

I would prefer adding the message-factory package as a dependency and inject it in the constructor.

**Comment by Nyholm** [comment]
Okey, Thank you for the feedback. I'll send a PR.

**Comment by Nyholm** [comment]
Closed by #21
